### PR TITLE
Add a `pyproject.toml` and minimal packaging structure

### LINKED ISSUE #5: Add a `pyproject.toml`
Package with Flit or Hatch.

### PR DESCRIPTION
**Pull request by FollowTheProcess**
Hey! 👋🏻

Saw this project come up on my GitHub feed and figured it's quick enough for me to help out with some of the issues you've put up.

This PR closes #5

I've put a fairly minimal `pyproject.toml` in using hatch. I've had to restructure the package slightly to make this work (I'm not sure if single file packages are possible, at least I've never done one). I've gone for `./src` layout but let me know if you'd like a different layout!

Also added `pytest` as a dev dependency with a config I've found useful in my projects. And added an entry point 🙂

Let me know if you'd like any changes or have any other comments 👍🏻

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+[project.optional-dependencies]
+dev = [
+  "pytest>=7.2.2",
+]
+
+[project.urls]
+source = "https://github.com/brettcannon/microvenv"
+
+[project.scripts]
+microvenv = "microvenv.main:main"
```
Comment (brettcannon) [inline pyproject.toml:46]: I [don't want an entry point](https://github.com/brettcannon/microvenv/issues/5#issuecomment-1455219067).

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+[tool.pytest.ini_options]
+minversion = "7.0"
+addopts = [ "-ra", "--strict-markers", "--strict-config" ]
+xfail_strict = true
+log_cli_level = "info"
+pythonpath = [ "src" ]
+testpaths = [ "tests" ]
```
Comment (brettcannon) [inline pyproject.toml:54]: What is all of this? Testing has not been set up yet.

**Review by brettcannon — CHANGES_REQUESTED** [review]
The project structure is unfortunately very purposeful, so I can't have this converted into a package; it needs to stay a self-contained file. As such, I don't think having a `src` layout is helpful and just makes running the tests a little bit harder by requiring an editable install for not benefit.

**Comment by brettcannon** [comment]
Thanks for the PR!

> I've put a fairly minimal `pyproject.toml` in using hatch. I've had to restructure the package slightly to make this work (I'm not sure if single file packages are possible, at least I've never done one). I've gone for `./src` layout but let me know if you'd like a different layout!

Unfortunately I need this to stay a single file.



> And added an entry point 🙂

Unfortunately I don't want an entry point (see the original issue as to why).

**Comment by brettcannon** [comment]
And FYI [Flit](https://flit.pypa.io/) supports a single file if Hatch doesn't.

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+readme = "README.md"
+license = "MIT"
+maintainers = [ { name = "Brett Cannon", email = "brett@python.org" } ]
+authors = [ { name = "Brett Cannon", email = "brett@python.org" } ]
+requires-python = ">=3.7"
+classifiers = [
+  "Development Status :: 3 - Alpha",
+  "Environment :: Console",
+  "Intended Audience :: Developers",
+  "License :: OSI Approved :: Apache Software License",
```
[the remark this answers is not in the document]
Comment (FollowTheProcess) [reply pyproject.toml:20]: Copy paste strikes again 🤦🏻‍♂️

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  "Operating System :: POSIX :: Linux",
+  "Operating System :: Microsoft :: Windows",
+  "Operating System :: OS Independent",
+  "Programming Language :: Python :: 3",
+  "Programming Language :: Python :: 3 :: Only",
+  "Programming Language :: Python :: 3.7",
+  "Programming Language :: Python :: 3.8",
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",
```
[the remark this answers is not in the document]
Comment (FollowTheProcess) [reply pyproject.toml:29]: I've kept this in as I've seen you've now got CI on all these set up

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+classifiers = [
+  "Development Status :: 3 - Alpha",
+  "Environment :: Console",
+  "Intended Audience :: Developers",
+  "License :: OSI Approved :: Apache Software License",
+  "Natural Language :: English",
+  "Operating System :: MacOS :: MacOS X",
+  "Operating System :: POSIX :: Linux",
+  "Operating System :: Microsoft :: Windows",
+  "Operating System :: OS Independent",
```
[the remark this answers is not in the document]
Comment (FollowTheProcess) [reply pyproject.toml]: Agreed, more copy paste. Got rid and just kept OS Independent 👍🏻

File: pyproject.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (FollowTheProcess) [reply pyproject.toml]: Fixed

File: pyproject.toml (hunk 1 above)
Comment (FollowTheProcess) [reply pyproject.toml]: Removed :)

File: pyproject.toml (hunk 2 above)
Comment (FollowTheProcess) [reply pyproject.toml]: Just some helpful pytest config I copied from one of my projects. Agree though it doesn't make much sense here as the project is so simple

**Comment by FollowTheProcess** [comment]
Looks like hatch supports a single file package out of the box, I did `hatch build` and unzipped the tarball it looks as if everything is there as expected.

Also noticed you've now added ruff as a linter. I have a fairly strict `ruff.toml` in my home directory that was making it trigger a load of warnings. I think just including the default ruff config in `pyproject.toml` should stop that happening for anyone else

Think I addressed all your comments, let me know if you want me to make any further changes 👍🏻

**Comment by brettcannon** [comment]
I marked all of the requested changes as resolved and refreshed my review. Unfortunately it looks like the changes you made to `microvenv.py` aren't formatted with Black. But since this is about to become a blocker for me releasing this project to PyPI, I'm going to override the failure and fix the code up myself.

Thanks for the PR!

**Pull request merged**